src/crypto/equihash.cpp

```diff
@@ -143,37 +143,73 @@ std::vector<eh_index> FullStepRow<WIDTH>::GetIndices(size_t len, size_t lenIndic
 }
 
 template<size_t WIDTH>
-bool HasCollision(StepRow<WIDTH>& a, StepRow<WIDTH>& b, int l)
+bool HasCollision(StepRow<WIDTH>& a, StepRow<WIDTH>& b, int o, int l)
 {
     // This doesn't need to be constant time.
     for (int j = 0; j < l; j++) {
-        if (a.hash[j] != b.hash[j])
+        if (a.hash[o+j] != b.hash[o+j])
             return false;
     }
     return true;
 }
 
 template<size_t WIDTH>
-TruncatedStepRow<WIDTH>::TruncatedStepRow(unsigned int n, const eh_HashState& base_state, eh_index i, unsigned int ilen) :
+TruncatedStepRow<WIDTH>::TruncatedStepRow(unsigned int n, const eh_HashState& base_state,
+                                          eh_index i, size_t ilen, size_t trim,
+                                          eh_index& bucket) :
         StepRow<WIDTH> {n, base_state, i}
 {
-    hash[n/8] = TruncateIndex(i, ilen);
+    assert(trim >= 1);
+    assert(trim <= sizeof(eh_index));
+    assert(1+n/8-trim+sizeof(eh_trunc) <= WIDTH);
+
+    // Bucket index is the first trim bytes of the hash
+    eh_index lei = 0;
+    memcpy(&lei, hash, trim);
+    bucket = le32toh(lei);
+
+    // Shift remaining hash bytes leaving space for the flag byte
+    if (trim > 1) {
+        for (int i = trim; i < n/8; i++)
+            hash[1+i-trim] = hash[i];
```

Comment:
It would be much better if we could generate the first `trim` bytes of the BLAKE2b hash directly into the bucket, and then the remaining `n/8 - trim` bytes into `hash`, instead of generating all `n/8` bytes into `hash` and then shifting `n/8 - trim` bytes over.